astropy/nddata/nddata.py

```diff
@@ -82,15 +82,15 @@ def __init__(self, data, uncertainty=None, mask=None, wcs=None,
 
         super(NDData, self).__init__()
 
-        if isinstance(data, self.__class__):
+        if isinstance(data, NDData):  # don't use self.__class__ (issue #4137)
             # No need to check the data because data must have successfully
             # initialized.
             self._data = data._data
+            self._unit = data.unit
```

Comment:
Could you add a comment here that `self._unit` needs to come before `self.uncertainty`? That should help avoid future edits that would break this.